---
title: "GENOVA distance-dependent decay and differences"
author: "Mikhail Dozmorov"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cerulean
    toc: yes
  pdf_document:
    toc: no
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
# Set up the environment
library(knitr)
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=F, cache.lazy = FALSE, tidy=T, fig.keep='high', echo=F, dpi=100, warnings=F, message=F, comment=NA, warning=F, results='as.is', fig.width = 10, fig.height = 6) #out.width=700, 
library(pander)
panderOptions('table.split.table', Inf)
set.seed(1)
library(dplyr)
options(stringsAsFactors = FALSE)
```

# Libraries

```{r libraries}
library(GENOVA) # remotes::install_github("robinweide/GENOVA", ref = "dev")
library(rCGH)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(strawr) # remotes::install_github("aidenlab/straw/R")
library(grid)
library(gridExtra)
library(ggplot2)
library(ggrepel)
library("ggsci")
library(scales)
# scales::show_col(pal_lancet("lanonc")(8))
mycols = pal_lancet("lanonc")(8)
library(grid)
library(gridExtra)
library(ggprism)
library(cowplot)
library(patchwork)
```

# Settings

```{r settings}
# .mcool files
fileNamemcool1 <- "/Users/mdozmorov/Documents/Data/GoogleDrive/HiC_files/03.Arima_samples_hic2cool/UCD52PR.mcool"
fileNamemcool2 <- "/Users/mdozmorov/Documents/Data/GoogleDrive/HiC_files/03.Arima_samples_hic2cool/UCD52CR.mcool"
# For saddle plot
nbins = 50 # Number of bins
```
```{r centromeres}
# Centromeres from rCGH
hg38_chrom <- hg38
hg38_chrom$chrom[hg38_chrom$chrom == 23] <- "X"
hg38_chrom$chrom[hg38_chrom$chrom == 24] <- "Y"
centromeres <- hg38_chrom[, c("chrom", "centromerStart", "centromerEnd")]
```

# Load data

```{r eval=TRUE}
UCD52PR_50kb_cool <- load_contacts(signal_path = fileNamemcool1,
                                   sample_name = "PR",
                                   resolution = 50e3,
                                   balancing = TRUE, # this is the default
                                   colour = "black",
                                   centromeres = centromeres)
UCD52CR_50kb_cool <- load_contacts(signal_path = fileNamemcool2,
                                   sample_name = "CR",
                                   resolution = 50e3,
                                   balancing = TRUE, # this is the default
                                   colour = "red",
                                   centromeres = centromeres)
```

```{r eval=TRUE}
UCD52PR_100kb_cool <- load_contacts(signal_path = fileNamemcool1,
                                   sample_name = "PR",
                                   resolution = 100e3,
                                   balancing = TRUE, # this is the default
                                   colour = "black",
                                   centromeres = centromeres)
UCD52CR_100kb_cool <- load_contacts(signal_path = fileNamemcool2,
                                   sample_name = "CR",
                                   resolution = 100e3,
                                   balancing = TRUE, # this is the default
                                   colour = "red",
                                   centromeres = centromeres)
```

```{r cooler}
UCD52PR_500kb_cool <- load_contacts(signal_path = fileNamemcool1, 
                                   sample_name = "PR", 
                                   resolution = 500e3, 
                                   balancing = TRUE, # this is the default
                                   colour = "black",
                                   centromeres = centromeres)
UCD52CR_500kb_cool <- load_contacts(signal_path = fileNamemcool2, 
                                   sample_name = "CR", 
                                   resolution = 500e3, 
                                   balancing = TRUE, # this is the default
                                   colour = "red",
                                   centromeres = centromeres)
```

## AB compartments

### 500kb resolution

```{r fig.height=12}
allGenes <- exons(TxDb.Hsapiens.UCSC.hg38.knownGene)
allGenes <- as.data.frame(allGenes)
allGenes$seqnames <- sub("chr", "", allGenes$seqnames)
allGenes <- allGenes[allGenes$seqnames %in% unique(UCD52PR_500kb_cool$CHRS) , 1:3]

CS_out = compartment_score(list(UCD52PR_500kb_cool, UCD52CR_500kb_cool), bed = allGenes)

# Look through each chromosome
gg_list <- list()
for (chrom in c(1:22, "X", "Y")) {
  gg <- visualise(CS_out, chr = chrom, title = chrom) 
  gg <- gg + theme(axis.title.y = element_blank(), legend.position = "none")
  # readline(prompt="Press [enter] to continue")
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```

#### Saddle-analyses

```{r fig.height=4}
UCD52PR_500kb_cool$CHRS <- UCD52PR_500kb_cool$CHRS[!(UCD52PR_500kb_cool$CHRS %in% c("M", "Y"))]
UCD52CR_500kb_cool$CHRS <- UCD52CR_500kb_cool$CHRS[!(UCD52CR_500kb_cool$CHRS %in% c("M", "Y"))]
saddle_out = saddle(list(UCD52PR_500kb_cool, UCD52CR_500kb_cool), 
                   CS_discovery = CS_out,
                   bins = nbins)

visualise(saddle_out)  + scale_altfill_continuous(low = mycols[1], high = mycols[2])
ggsave("../manuscript/figures/Figure_saddle_500kb.svg", width = 7, height = 7)
```

```{r fig.height=3}
# # Look through each chromosome
# for (chrom in c(1:22, "X", "Y")) {
#   p <- visualise(saddle_out, chr = chrom, title = chrom) + scale_altfill_continuous(low = mycols[1], high = mycols[2])
#   print(p)
#   # readline(prompt="Press [enter] to continue")
# }
```

#### Compartment-strength

```{r saddleStrength500, fig.height = 4, fig.width = 4}
CSS <- quantify(saddle_out)
# Exclude short arms where calculations produce outliers
# Check for outliers
# ggplot(CSS, aes(x = exp, y = strength)) +
#   geom_boxplot() +
#   geom_label_repel(aes(label = chr))
# Manually exclude them
# CSS <- CSS[!(CSS$chr %in% c("21p", "22p")), ]

compared <- tidyr::spread(unique(CSS[,-c(3,4)]), key = 'exp', value = 'strength')
# Compare compartmentalization strength
print(paste("Mean compartment strength in PR:", mean(compared$PR)))
print(paste("Mean compartment strength in CR:", mean(compared$CR)))
print(paste("T-test:", t.test(compared$PR, compared$CR)$p.value %>% formatC(., format = "f", digits = 3)))

# Plot compartment strength differences
ggplot(compared, aes(x = PR, y = CR)) +
  geom_segment(aes(x = 1, xend = 4, y = 1, yend = 4, colour = "grey")) + scale_colour_identity() +
  geom_point(col = "red") +
  geom_label_repel(aes(label = chr), force=1, box.padding=0.5, label.padding = 0.1, segment.color = 'grey50', max.overlaps = 5) +
  theme_bw() + xlim(1, 4) + ylim(1, 4) +
  labs(title = "Compartment strength comparison")
```

#### Du_et_al_R_processing

##### Saddle plot, genome-wide 

```{r fig.height=3}
#--- Figure 4A - saddle plot ----
# saddle_out from GENOVA
# head(saddle_out$saddle)
dat <- saddle_out$saddle
dat2 <- dat[which(dat$mean > 0),]

tmp = setNames(aggregate(log2(dat2$mean), by = list(dat2$q1, dat2$q2, dat2$exp), mean, na.rm = T), c("x",'y','z', "score"))
tmp$score[which(tmp$score < -1)] <- -1
tmp$score[which(tmp$score > 1)] <- 1

comp <- tmp[tmp$x != tmp$y,]
colnames(comp) <- c("y", "x", "z", "score")
comp <- comp[,c("x", "y", "z", "score")]

df <- rbind(tmp, comp)
df$z <- factor(df$z, levels=c("PR", "CR"))

gg <- ggplot(df, aes(x, y, fill = score)) + geom_tile() + scale_x_reverse() +
  scale_fill_gradientn(colours=rev(c("#B2182B", "white", "#2166AC")), limits=c(-1, 1)) + theme_minimal() + 
  theme(panel.grid = element_blank()) + ggtitle("saddle") +
  coord_fixed() + labs(x="bin1", y="bin2") + facet_wrap(~z) # axis.text.x = element_text(angle=90, vjust = 1, hjust = 1), 
gg
ggsave("../manuscript/figures/Figure_saddle_500kb_Du.svg", width = 7, height = 3.5)
```

##### Contact enrichment, genome-wide

```{r fig.height=2, fig.width=4}
#--- Figure 4B ----
# saddle_out from GENOVA
# script from GENOVA compartment strength function
dat = saddle_out$saddle
dat <- dat[!is.na(dat$q1),]

namesVector <- c()
namesVector <- c(namesVector, unique(dat$exp))
dat$CC <- 'XX'
MAXbin = max(dat$q1)
binsTOse = floor(MAXbin * .2)
binsTOse = max(1, binsTOse)
#dat$MAT$unLog = 2 ** dat$MAT$value #values are already unlog-ed
dat[dat$q1 <= binsTOse & dat$q2 <= binsTOse,"CC"] = "BB"
dat[dat$q2 <= binsTOse & dat$q1 >= MAXbin-binsTOse+1,"CC"] = "AB"
dat[dat$q1 <= binsTOse & dat$q2 >= MAXbin-binsTOse+1,"CC"] = "BA"
dat[dat$q1 >= MAXbin-binsTOse+1 & dat$q2 >= MAXbin-binsTOse+1,"CC"] = "AA"
dat = dat[dat$CC != 'XX',]
dat2 <- dat[which(dat$mean > 0),]

cescores = dplyr::summarise(dplyr::group_by(dat2,
                                       exp,
                                       CC),score = mean(log2(mean)))
## values of cescores are plotted in 4B
cescores$exp <- factor(cescores$exp, levels = c("PR", "CR"))
cescores$CC  <- factor(cescores$CC, levels = c("AA", "BB", "BA"))
ggplot(cescores, aes(x = exp, y = score, fill = exp)) +
  geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
  scale_fill_manual(values = mycols[1:2]) +
  theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
  facet_wrap(.~CC, scales = "free")
ggsave("../manuscript/figures/Figure_contact_enrichment_500kb_Du.svg", width = 5.5, height = 3)
```

##### Contact enrichment, per chromosome

```{r fig.height=20}
dat = saddle_out$saddle
dat <- dat[!is.na(dat$q1),]
# Sort chromosome arms in logical order
chromosomes <- unique(dat$chr) %>% gtools::mixedsort()
gg_list <- list()
for (chrom in chromosomes) {
  dat_selected <- dat %>% filter(chr == chrom)

  namesVector <- c()
  namesVector <- c(namesVector, unique(dat_selected$exp))
  dat_selected$CC <- 'XX'
  MAXbin = max(dat_selected$q1)
  binsTOse = floor(MAXbin * .2)
  binsTOse = max(1, binsTOse)
  #dat_selected$MAT$unLog = 2 ** dat_selected$MAT$value #values are already unlog-ed
  dat_selected[dat_selected$q1 <= binsTOse & dat_selected$q2 <= binsTOse,"CC"] = "BB"
  dat_selected[dat_selected$q2 <= binsTOse & dat_selected$q1 >= MAXbin-binsTOse+1,"CC"] = "AB"
  dat_selected[dat_selected$q1 <= binsTOse & dat_selected$q2 >= MAXbin-binsTOse+1,"CC"] = "BA"
  dat_selected[dat_selected$q1 >= MAXbin-binsTOse+1 & dat_selected$q2 >= MAXbin-binsTOse+1,"CC"] = "AA"
  dat_selected = dat_selected[dat_selected$CC != 'XX',]
  dat_selected2 <- dat_selected[which(dat_selected$mean > 0),]
  cescores_selected = dplyr::summarise(dplyr::group_by(dat_selected2,
                                              exp,
                                              CC),score = mean(log2(mean)))
  ## values of cescores_selected are plotted in 4B
  cescores_selected$exp <- factor(cescores_selected$exp, levels = c("PR", "CR"))
  cescores_selected$CC  <- factor(cescores_selected$CC, levels = c("AA", "BB", "BA"))
  gg <- ggplot(cescores_selected, aes(x = CC, y = score, fill = exp)) +
    geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.9)) +
    scale_fill_manual(values = mycols[1:2]) +
    theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
    ggtitle(chrom)
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```

##### Compartmentalization score, genome-wide

```{r fig.height=2, fig.width=3}
#--- Figure 4C ----
tmp = dplyr::summarise(dplyr::group_by(dat2, chr,
                                       exp,
                                       CC),score = mean(mean))
df.tmp <- as.data.frame(tmp)
df.tmp <- reshape2::dcast(df.tmp, exp + chr ~ CC)
df.tmp$compscore <- log((df.tmp$AA*df.tmp$BB)/(df.tmp$BA^2))

df.tmp$exp <- factor(df.tmp$exp, levels=c("PR", "CR"))

df.hd <- dplyr::summarise(dplyr::group_by(df.tmp[is.finite(df.tmp$compscore),], exp), score = mean(compscore))
## values of df.hd are plotted in 4B
df.hd$exp <- factor(df.hd$exp, levels = c("PR", "CR"))
ggplot(df.hd, aes(x = exp, y = score, fill = exp)) +
  geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
  scale_fill_manual(values = mycols[1:2]) +
  theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) 
ggsave("../manuscript/figures/Figure_compartmentalization_score_500kb_Du.svg", width = 2.5, height = 3)
```

##### Compartmentalization score, chromosome-specific

```{r fig.height=20}
#--- Figure 4C ----
tmp = dplyr::summarise(dplyr::group_by(dat2, chr,
                                       exp,
                                       CC),score = mean(mean))
df.tmp <- as.data.frame(tmp)
# Sort chromosome arms in logical order
chromosomes <- unique(dat$chr) %>% gtools::mixedsort()
gg_list <- list()
for (chrom in chromosomes) {
  df.tmp_selected <- df.tmp %>% filter(chr == chrom)
  df.tmp_selected <- reshape2::dcast(df.tmp_selected, exp + chr ~ CC)
  df.tmp_selected$compscore <- log((df.tmp_selected$AA*df.tmp_selected$BB)/(df.tmp_selected$BA^2))
  
  df.tmp_selected$exp <- factor(df.tmp_selected$exp, levels=c("PR", "CR"))
  
  df.hd_selected <- dplyr::summarise(dplyr::group_by(df.tmp_selected[is.finite(df.tmp_selected$compscore),], exp), score = mean(compscore))
  ## values of df.hd_selected are plotted in 4B
  df.hd_selected$exp <- factor(df.hd_selected$exp, levels = c("PR", "CR"))
  gg <- ggplot(df.hd_selected, aes(x = exp, y = score, fill = exp)) +
    geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
    scale_fill_manual(values = mycols[1:2]) +
    theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
    ggtitle(chrom)
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```


### 100kb resolution

```{r fig.height=12}
allGenes <- exons(TxDb.Hsapiens.UCSC.hg38.knownGene)
allGenes <- as.data.frame(allGenes)
allGenes$seqnames <- sub("chr", "", allGenes$seqnames)
allGenes <- allGenes[allGenes$seqnames %in% unique(UCD52PR_100kb_cool$CHRS) , 1:3]

CS_out = compartment_score(list(UCD52PR_100kb_cool, UCD52CR_100kb_cool), bed = allGenes)

# Look through each chromosome
gg_list <- list()
for (chrom in c(1:22, "X", "Y")) {
  gg <- visualise(CS_out, chr = chrom, title = chrom) 
  gg <- gg + theme(axis.title.y = element_blank(), legend.position = "none")
  # readline(prompt="Press [enter] to continue")
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```

#### Saddle-analyses

```{r fig.height=4}
UCD52PR_100kb_cool$CHRS <- UCD52PR_100kb_cool$CHRS[!(UCD52PR_100kb_cool$CHRS %in% c("M", "Y"))]
UCD52CR_100kb_cool$CHRS <- UCD52CR_100kb_cool$CHRS[!(UCD52CR_100kb_cool$CHRS %in% c("M", "Y"))]
saddle_out = saddle(list(UCD52PR_100kb_cool, UCD52CR_100kb_cool), 
                   CS_discovery = CS_out,
                   bins = nbins)

visualise(saddle_out)  + scale_altfill_continuous(low = mycols[1], high = mycols[2])
ggsave("../manuscript/figures/Figure_saddle_100kb.svg", width = 7, height = 7)
```

```{r fig.height=3}
# # Look through each chromosome
# for (chrom in c(1:22, "X", "Y")) {
#   p <- visualise(saddle_out, chr = chrom, title = chrom) + scale_altfill_continuous(low = mycols[1], high = mycols[2])
#   print(p)
#   # readline(prompt="Press [enter] to continue")
# }
```

#### Compartment-strength

```{r saddleStrength100, fig.height = 4, fig.width = 4}
CSS <- quantify(saddle_out)
# Exclude short arms where calculations produce outliers
# Check for outliers
# ggplot(CSS, aes(x = exp, y = strength)) +
#   geom_boxplot() +
#   geom_label_repel(aes(label = chr))
# Manually exclude them
# CSS <- CSS[!(CSS$chr %in% c("21p", "22p")), ]

compared <- tidyr::spread(unique(CSS[,-c(3,4)]), key = 'exp', value = 'strength')
# Compare compartmentalization strength
print(paste("Mean compartment strength in PR:", mean(compared$PR)))
print(paste("Mean compartment strength in CR:", mean(compared$CR)))
print(paste("T-test:", t.test(compared$PR, compared$CR)$p.value %>% formatC(., format = "f", digits = 3)))

# Plot compartment strength differences
ggplot(compared, aes(x = PR, y = CR)) +
  geom_segment(aes(x = 1, xend = 4, y = 1, yend = 4, colour = "grey")) + scale_colour_identity() +
  geom_point(col = "red") +
  geom_label_repel(aes(label = chr), force=1, box.padding=0.5, label.padding = 0.1, segment.color = 'grey50', max.overlaps = 5) +
  theme_bw() + xlim(1, 4) + ylim(1, 4) +
  labs(title = "Compartment strength comparison")
```

#### Du_et_al_R_processing

##### Saddle plot, genome-wide 

```{r fig.height=3}
#--- Figure 4A - saddle plot ----
# saddle_out from GENOVA
# head(saddle_out$saddle)
dat <- saddle_out$saddle
dat2 <- dat[which(dat$mean > 0),]

tmp = setNames(aggregate(log2(dat2$mean), by = list(dat2$q1, dat2$q2, dat2$exp), mean, na.rm = T), c("x",'y','z', "score"))
tmp$score[which(tmp$score < -1)] <- -1
tmp$score[which(tmp$score > 1)] <- 1

comp <- tmp[tmp$x != tmp$y,]
colnames(comp) <- c("y", "x", "z", "score")
comp <- comp[,c("x", "y", "z", "score")]

df <- rbind(tmp, comp)
df$z <- factor(df$z, levels=c("PR", "CR"))

gg <- ggplot(df, aes(x, y, fill = score)) + geom_tile() + scale_x_reverse() +
  scale_fill_gradientn(colours=rev(c("#B2182B", "white", "#2166AC")), limits=c(-1, 1)) + theme_minimal() + 
  theme(panel.grid = element_blank()) + ggtitle("saddle") +
  coord_fixed() + labs(x="bin1", y="bin2") + facet_wrap(~z) # axis.text.x = element_text(angle=90, vjust = 1, hjust = 1), 
gg
ggsave("../manuscript/figures/Figure_saddle_100kb_Du.svg", width = 7, height = 3.5)
```

##### Contact enrichment, genome-wide

```{r fig.height=2, fig.width=4}
#--- Figure 4B ----
# saddle_out from GENOVA
# script from GENOVA compartment strength function
dat = saddle_out$saddle
dat <- dat[!is.na(dat$q1),]

namesVector <- c()
namesVector <- c(namesVector, unique(dat$exp))
dat$CC <- 'XX'
MAXbin = max(dat$q1)
binsTOse = floor(MAXbin * .2)
binsTOse = max(1, binsTOse)
#dat$MAT$unLog = 2 ** dat$MAT$value #values are already unlog-ed
dat[dat$q1 <= binsTOse & dat$q2 <= binsTOse,"CC"] = "BB"
dat[dat$q2 <= binsTOse & dat$q1 >= MAXbin-binsTOse+1,"CC"] = "AB"
dat[dat$q1 <= binsTOse & dat$q2 >= MAXbin-binsTOse+1,"CC"] = "BA"
dat[dat$q1 >= MAXbin-binsTOse+1 & dat$q2 >= MAXbin-binsTOse+1,"CC"] = "AA"
dat = dat[dat$CC != 'XX',]
dat2 <- dat[which(dat$mean > 0),]

cescores = dplyr::summarise(dplyr::group_by(dat2,
                                       exp,
                                       CC),score = mean(log2(mean)))
## values of cescores are plotted in 4B
cescores$exp <- factor(cescores$exp, levels = c("PR", "CR"))
cescores$CC  <- factor(cescores$CC, levels = c("AA", "BB", "BA"))
ggplot(cescores, aes(x = exp, y = score, fill = exp)) +
  geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
  scale_fill_manual(values = mycols[1:2]) +
  theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
  facet_wrap(.~CC, scales = "free")
ggsave("../manuscript/figures/Figure_contact_enrichment_100kb_Du.svg", width = 5.5, height = 3)
```

##### Contact enrichment, per chromosome

```{r fig.height=20}
dat = saddle_out$saddle
dat <- dat[!is.na(dat$q1),]
# Sort chromosome arms in logical order
chromosomes <- unique(dat$chr) %>% gtools::mixedsort()
gg_list <- list()
for (chrom in chromosomes) {
  dat_selected <- dat %>% filter(chr == chrom)

  namesVector <- c()
  namesVector <- c(namesVector, unique(dat_selected$exp))
  dat_selected$CC <- 'XX'
  MAXbin = max(dat_selected$q1)
  binsTOse = floor(MAXbin * .2)
  binsTOse = max(1, binsTOse)
  #dat_selected$MAT$unLog = 2 ** dat_selected$MAT$value #values are already unlog-ed
  dat_selected[dat_selected$q1 <= binsTOse & dat_selected$q2 <= binsTOse,"CC"] = "BB"
  dat_selected[dat_selected$q2 <= binsTOse & dat_selected$q1 >= MAXbin-binsTOse+1,"CC"] = "AB"
  dat_selected[dat_selected$q1 <= binsTOse & dat_selected$q2 >= MAXbin-binsTOse+1,"CC"] = "BA"
  dat_selected[dat_selected$q1 >= MAXbin-binsTOse+1 & dat_selected$q2 >= MAXbin-binsTOse+1,"CC"] = "AA"
  dat_selected = dat_selected[dat_selected$CC != 'XX',]
  dat_selected2 <- dat_selected[which(dat_selected$mean > 0),]
  cescores_selected = dplyr::summarise(dplyr::group_by(dat_selected2,
                                              exp,
                                              CC),score = mean(log2(mean)))
  ## values of cescores_selected are plotted in 4B
  cescores_selected$exp <- factor(cescores_selected$exp, levels = c("PR", "CR"))
  cescores_selected$CC  <- factor(cescores_selected$CC, levels = c("AA", "BB", "BA"))
  gg <- ggplot(cescores_selected, aes(x = CC, y = score, fill = exp)) +
    geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.9)) +
    scale_fill_manual(values = mycols[1:2]) +
    theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
    ggtitle(chrom)
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```

##### Compartmentalization score, genome-wide

```{r fig.height=2, fig.width=3}
#--- Figure 4C ----
tmp = dplyr::summarise(dplyr::group_by(dat2, chr,
                                       exp,
                                       CC),score = mean(mean))
df.tmp <- as.data.frame(tmp)
df.tmp <- reshape2::dcast(df.tmp, exp + chr ~ CC)
df.tmp$compscore <- log((df.tmp$AA*df.tmp$BB)/(df.tmp$BA^2))

df.tmp$exp <- factor(df.tmp$exp, levels=c("PR", "CR"))

df.hd <- dplyr::summarise(dplyr::group_by(df.tmp[is.finite(df.tmp$compscore),], exp), score = mean(compscore))
## values of df.hd are plotted in 4B
df.hd$exp <- factor(df.hd$exp, levels = c("PR", "CR"))
ggplot(df.hd, aes(x = exp, y = score, fill = exp)) +
  geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
  scale_fill_manual(values = mycols[1:2]) +
  theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) 
ggsave("../manuscript/figures/Figure_compartmentalization_score_100kb_Du.svg", width = 2.5, height = 3)
```

##### Compartmentalization score, chromosome-specific

```{r fig.height=20}
#--- Figure 4C ----
tmp = dplyr::summarise(dplyr::group_by(dat2, chr,
                                       exp,
                                       CC),score = mean(mean))
df.tmp <- as.data.frame(tmp)
# Sort chromosome arms in logical order
chromosomes <- unique(dat$chr) %>% gtools::mixedsort()
gg_list <- list()
for (chrom in chromosomes) {
  df.tmp_selected <- df.tmp %>% filter(chr == chrom)
  df.tmp_selected <- reshape2::dcast(df.tmp_selected, exp + chr ~ CC)
  df.tmp_selected$compscore <- log((df.tmp_selected$AA*df.tmp_selected$BB)/(df.tmp_selected$BA^2))
  
  df.tmp_selected$exp <- factor(df.tmp_selected$exp, levels=c("PR", "CR"))
  
  df.hd_selected <- dplyr::summarise(dplyr::group_by(df.tmp_selected[is.finite(df.tmp_selected$compscore),], exp), score = mean(compscore))
  ## values of df.hd_selected are plotted in 4B
  df.hd_selected$exp <- factor(df.hd_selected$exp, levels = c("PR", "CR"))
  gg <- ggplot(df.hd_selected, aes(x = exp, y = score, fill = exp)) +
    geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
    scale_fill_manual(values = mycols[1:2]) +
    theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
    ggtitle(chrom)
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```

### 50kb resolution

```{r fig.height=12}
allGenes <- exons(TxDb.Hsapiens.UCSC.hg38.knownGene)
allGenes <- as.data.frame(allGenes)
allGenes$seqnames <- sub("chr", "", allGenes$seqnames)
allGenes <- allGenes[allGenes$seqnames %in% unique(UCD52PR_50kb_cool$CHRS) , 1:3]

CS_out = compartment_score(list(UCD52PR_50kb_cool, UCD52CR_50kb_cool), bed = allGenes)

# Look through each chromosome
gg_list <- list()
for (chrom in c(1:22, "X", "Y")) {
  gg <- visualise(CS_out, chr = chrom, title = chrom) 
  gg <- gg + theme(axis.title.y = element_blank(), legend.position = "none")
  # readline(prompt="Press [enter] to continue")
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```

#### Saddle-analyses

```{r fig.height=4}
UCD52PR_50kb_cool$CHRS <- UCD52PR_50kb_cool$CHRS[!(UCD52PR_50kb_cool$CHRS %in% c("M", "Y"))]
UCD52CR_50kb_cool$CHRS <- UCD52CR_50kb_cool$CHRS[!(UCD52CR_50kb_cool$CHRS %in% c("M", "Y"))]
saddle_out = saddle(list(UCD52PR_50kb_cool, UCD52CR_50kb_cool), 
                   CS_discovery = CS_out,
                   bins = nbins)

visualise(saddle_out)  + scale_altfill_continuous(low = mycols[1], high = mycols[2])
ggsave("../manuscript/figures/Figure_saddle_50kb.svg", width = 7, height = 7)
```

```{r fig.height=3}
# # Look through each chromosome
# for (chrom in c(1:22, "X", "Y")) {
#   p <- visualise(saddle_out, chr = chrom, title = chrom) + scale_altfill_continuous(low = mycols[1], high = mycols[2])
#   print(p)
#   # readline(prompt="Press [enter] to continue")
# }
```

#### Compartment-strength

```{r saddleStrength50, fig.height = 4, fig.width = 4}
CSS <- quantify(saddle_out)
# Exclude short arms where calculations produce outliers
# Check for outliers
# ggplot(CSS, aes(x = exp, y = strength)) +
#   geom_boxplot() +
#   geom_label_repel(aes(label = chr))
# Manually exclude them
# CSS <- CSS[!(CSS$chr %in% c("21p", "22p")), ]

compared <- tidyr::spread(unique(CSS[,-c(3,4)]), key = 'exp', value = 'strength')
# Compare compartmentalization strength
print(paste("Mean compartment strength in PR:", mean(compared$PR)))
print(paste("Mean compartment strength in CR:", mean(compared$CR)))
print(paste("T-test:", t.test(compared$PR, compared$CR)$p.value %>% formatC(., format = "f", digits = 3)))

# Plot compartment strength differences
ggplot(compared, aes(x = PR, y = CR)) +
  geom_segment(aes(x = 1, xend = 4, y = 1, yend = 4, colour = "grey")) + scale_colour_identity() +
  geom_point(col = "red") +
  geom_label_repel(aes(label = chr), force=1, box.padding=0.5, label.padding = 0.1, segment.color = 'grey50', max.overlaps = 5) +
  theme_bw() + xlim(1, 4) + ylim(1, 4) +
  labs(title = "Compartment strength comparison")
```

#### Du_et_al_R_processing

##### Saddle plot, genome-wide 

```{r fig.height=3}
#--- Figure 4A - saddle plot ----
# saddle_out from GENOVA
# head(saddle_out$saddle)
dat <- saddle_out$saddle
dat2 <- dat[which(dat$mean > 0),]

tmp = setNames(aggregate(log2(dat2$mean), by = list(dat2$q1, dat2$q2, dat2$exp), mean, na.rm = T), c("x",'y','z', "score"))
tmp$score[which(tmp$score < -1)] <- -1
tmp$score[which(tmp$score > 1)] <- 1

comp <- tmp[tmp$x != tmp$y,]
colnames(comp) <- c("y", "x", "z", "score")
comp <- comp[,c("x", "y", "z", "score")]

df <- rbind(tmp, comp)
df$z <- factor(df$z, levels=c("PR", "CR"))

gg <- ggplot(df, aes(x, y, fill = score)) + geom_tile() + scale_x_reverse() +
  scale_fill_gradientn(colours=rev(c("#B2182B", "white", "#2166AC")), limits=c(-1, 1)) + theme_minimal() + 
  theme(panel.grid = element_blank()) + ggtitle("saddle") +
  coord_fixed() + labs(x="bin1", y="bin2") + facet_wrap(~z) # axis.text.x = element_text(angle=90, vjust = 1, hjust = 1), 
gg
ggsave("../manuscript/figures/Figure_saddle_50kb_Du.svg", width = 7, height = 3.5)
```

##### Contact enrichment, genome-wide

```{r fig.height=2, fig.width=4}
#--- Figure 4B ----
# saddle_out from GENOVA
# script from GENOVA compartment strength function
dat = saddle_out$saddle
dat <- dat[!is.na(dat$q1),]

namesVector <- c()
namesVector <- c(namesVector, unique(dat$exp))
dat$CC <- 'XX'
MAXbin = max(dat$q1)
binsTOse = floor(MAXbin * .2)
binsTOse = max(1, binsTOse)
#dat$MAT$unLog = 2 ** dat$MAT$value #values are already unlog-ed
dat[dat$q1 <= binsTOse & dat$q2 <= binsTOse,"CC"] = "BB"
dat[dat$q2 <= binsTOse & dat$q1 >= MAXbin-binsTOse+1,"CC"] = "AB"
dat[dat$q1 <= binsTOse & dat$q2 >= MAXbin-binsTOse+1,"CC"] = "BA"
dat[dat$q1 >= MAXbin-binsTOse+1 & dat$q2 >= MAXbin-binsTOse+1,"CC"] = "AA"
dat = dat[dat$CC != 'XX',]
dat2 <- dat[which(dat$mean > 0),]

cescores = dplyr::summarise(dplyr::group_by(dat2,
                                       exp,
                                       CC),score = mean(log2(mean)))
## values of cescores are plotted in 4B
cescores$exp <- factor(cescores$exp, levels = c("PR", "CR"))
cescores$CC  <- factor(cescores$CC, levels = c("AA", "BB", "BA"))
ggplot(cescores, aes(x = exp, y = score, fill = exp)) +
  geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
  scale_fill_manual(values = mycols[1:2]) +
  theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
  facet_wrap(.~CC, scales = "free")
ggsave("../manuscript/figures/Figure_contact_enrichment_50kb_Du.svg", width = 5.5, height = 3)
```

##### Contact enrichment, per chromosome

```{r fig.height=20}
dat = saddle_out$saddle
dat <- dat[!is.na(dat$q1),]
# Sort chromosome arms in logical order
chromosomes <- unique(dat$chr) %>% gtools::mixedsort()
gg_list <- list()
for (chrom in chromosomes) {
  dat_selected <- dat %>% filter(chr == chrom)

  namesVector <- c()
  namesVector <- c(namesVector, unique(dat_selected$exp))
  dat_selected$CC <- 'XX'
  MAXbin = max(dat_selected$q1)
  binsTOse = floor(MAXbin * .2)
  binsTOse = max(1, binsTOse)
  #dat_selected$MAT$unLog = 2 ** dat_selected$MAT$value #values are already unlog-ed
  dat_selected[dat_selected$q1 <= binsTOse & dat_selected$q2 <= binsTOse,"CC"] = "BB"
  dat_selected[dat_selected$q2 <= binsTOse & dat_selected$q1 >= MAXbin-binsTOse+1,"CC"] = "AB"
  dat_selected[dat_selected$q1 <= binsTOse & dat_selected$q2 >= MAXbin-binsTOse+1,"CC"] = "BA"
  dat_selected[dat_selected$q1 >= MAXbin-binsTOse+1 & dat_selected$q2 >= MAXbin-binsTOse+1,"CC"] = "AA"
  dat_selected = dat_selected[dat_selected$CC != 'XX',]
  dat_selected2 <- dat_selected[which(dat_selected$mean > 0),]
  cescores_selected = dplyr::summarise(dplyr::group_by(dat_selected2,
                                              exp,
                                              CC),score = mean(log2(mean)))
  ## values of cescores_selected are plotted in 4B
  cescores_selected$exp <- factor(cescores_selected$exp, levels = c("PR", "CR"))
  cescores_selected$CC  <- factor(cescores_selected$CC, levels = c("AA", "BB", "BA"))
  gg <- ggplot(cescores_selected, aes(x = CC, y = score, fill = exp)) +
    geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.9)) +
    scale_fill_manual(values = mycols[1:2]) +
    theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
    ggtitle(chrom)
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```

##### Compartmentalization score, genome-wide

```{r fig.height=2, fig.width=3}
#--- Figure 4C ----
tmp = dplyr::summarise(dplyr::group_by(dat2, chr,
                                       exp,
                                       CC),score = mean(mean))
df.tmp <- as.data.frame(tmp)
df.tmp <- reshape2::dcast(df.tmp, exp + chr ~ CC)
df.tmp$compscore <- log((df.tmp$AA*df.tmp$BB)/(df.tmp$BA^2))

df.tmp$exp <- factor(df.tmp$exp, levels=c("PR", "CR"))

df.hd <- dplyr::summarise(dplyr::group_by(df.tmp[is.finite(df.tmp$compscore),], exp), score = mean(compscore))
## values of df.hd are plotted in 4B
df.hd$exp <- factor(df.hd$exp, levels = c("PR", "CR"))
ggplot(df.hd, aes(x = exp, y = score, fill = exp)) +
  geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
  scale_fill_manual(values = mycols[1:2]) +
  theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) 
ggsave("../manuscript/figures/Figure_compartmentalization_score_50kb_Du.svg", width = 2.5, height = 3)
```

##### Compartmentalization score, chromosome-specific

```{r fig.height=20}
#--- Figure 4C ----
tmp = dplyr::summarise(dplyr::group_by(dat2, chr,
                                       exp,
                                       CC),score = mean(mean))
df.tmp <- as.data.frame(tmp)
# Sort chromosome arms in logical order
chromosomes <- unique(dat$chr) %>% gtools::mixedsort()
gg_list <- list()
for (chrom in chromosomes) {
  df.tmp_selected <- df.tmp %>% filter(chr == chrom)
  df.tmp_selected <- reshape2::dcast(df.tmp_selected, exp + chr ~ CC)
  df.tmp_selected$compscore <- log((df.tmp_selected$AA*df.tmp_selected$BB)/(df.tmp_selected$BA^2))
  
  df.tmp_selected$exp <- factor(df.tmp_selected$exp, levels=c("PR", "CR"))
  
  df.hd_selected <- dplyr::summarise(dplyr::group_by(df.tmp_selected[is.finite(df.tmp_selected$compscore),], exp), score = mean(compscore))
  ## values of df.hd_selected are plotted in 4B
  df.hd_selected$exp <- factor(df.hd_selected$exp, levels = c("PR", "CR"))
  gg <- ggplot(df.hd_selected, aes(x = exp, y = score, fill = exp)) +
    geom_bar(stat = "identity", width = 0.8, position = position_dodge(width = 0.5)) +
    scale_fill_manual(values = mycols[1:2]) +
    theme_classic() + theme(axis.line.x=element_blank(), axis.text.y = element_text(size=10)) +
    ggtitle(chrom)
  gg_list <- c(gg_list, list(gg))
}
wrap_plots(gg_list, ncol = 4)
```

