# Functions and transcription factors enriched in condition-specific loop anchors

<!-- Desktop/PDXHiC/Maggie/02_Resolution_Separated_Gene_Annotations.Rmd -->
- `01_Anchor_Gene_Annotations.Rmd` - annotating loop anchors by overlapping genes
  - Input: loop anchor bed files for each condition, resolution, and data type
  - Output: anchor bed files annotated with overlapping genes in comma-separated format
  
<!-- Desktop/PDXHiC/Maggie/03_Resolution_Separated_Gene_Enrichment.Rmd -->
- `02_Anchor_Gene_Enrichment.Rmd` - KEGG & MSigDb hypergeometric enrichment analyses done on loop anchors 
  - Input: annotated loop anchor bed files generated by `01_Anchor_Gene_Annotations.Rmd`
  - Output: `Supplementary_Table_4.xlsx` KEGG & MSigDb enrichment results for each condition

<!-- Desktop/PDXHiC/Maggie/41_Mustache_10kb_Enrichment_Barplots.Rmd -->
- `03_Anchors_Enrichment_Barplots.Rmd` - significant hypergeometric enrichments of genes overlapping condition-specific loop anchors in KEGG pathways
  - Input: `Supplementary_Table_4.xlsx` result from performing hypergeometric KEGG enrichment on genes overlapping 10kb Mustache anchors in PR and CR conditions 
  - Output: grouped barplot of -log10(p.adj) values for PR-unique and CR-unique conditions in the top KEGG pathways 
  
<!-- Desktop/PDXHiC/Maggie/05_Anchor_Masking.Rmd -->
- `04a_Anchor_Masking.Rmd` - masking loop anchors with open chromatin regions (ATAC-seq) 
  - Input: loop anchor bed files for each condition, resolution, and data type
  - Output: anchor bed files masked to contain only open chromatin regions
  
<!-- Desktop/PDXHiC/Maggie/06b_JASPAR_HOCOMOCO_MEME_IA.Rmd -->
- `04b_JASPAR_MEME_IA.Rmd` - MEME motif enrichment analysis for loop anchors at 10kb resolution
  - Input: loop anchor bed files masked with open chromatin regions produced by `04a_Anchor_Masking.Rmd`
  - Output: MEME motif enrichment heatmap for each type of loop anchor (Common/Unique) and condition (PR/CR) 
  
<!-- Desktop/PDXHiC/Mikhail/08_UniBind_plots.Rmd -->
- `05_Unibind_Plots.Rmd` - UniBind beeswarm plots for 10kb and 25kb loop anchors
  - Input: output from running UniBind differential analysis on masked CR loop anchors vs. masked PR loop anchors produced by `04a_Anchor_Masking.Rmd`
  - `allEnrichments.tsv` - UniBind results (source: /Users/mdozmorov/Documents/Data/GoogleDrive/HiC_files/results/Maggie/Mustache_results_v2/preprocessing_any/Motif_Enrichment/UniBind_results/Differential_No_Background/CR_vs_PR/Diff_10kb_CR_vs_PR/allEnrichments.tsv)
  - Output: UniBind transcription factor enrichment plots for each resolution 
  
<!-- Desktop/PDXHiC/Maggie/42_Supplementary_Gene_Enrichment_Barplots.Rmd -->
- `06_Integrative_Gene_Barplots.Rmd` - barplots of hypergeometric KEGG & MSigDb enrichments done on important genes in PR or CR conditions 
  - Input: `Supplementary_Table_6.xlsx` an excel file containing results from hypergeometric KEGG & MSigDb enrichments done on genes supported by 4 or more pieces of evidence in PR/CR conditions 
  - Output: barplot showing -log10(p.adj) value for the top enriched KEGG & MSigDb pathways in each condition 
